src/libcrun/cgroup.c

```diff
@@ -736,7 +738,71 @@ systemd_job_removed (sd_bus_message *m, void *userdata, sd_bus_error *error arg_
     return -1;
 
   if (strcmp (p->path, path) == 0)
-    *p->terminated = 1;
+    {
+      p->terminated = 1;
+      p->unit = xstrdup(unit);
+      p->result = xstrdup(result);
+    }
+  return 0;
+}
+
+static int
+systemd_check_job_status_setup (sd_bus *bus,
+                                void **data,
+                                libcrun_error_t *err)
+{
+  int ret;
+
+  *data = xmalloc(sizeof(struct systemd_job_removed_s));
+  ret = sd_bus_match_signal (bus,
+                             NULL,
+                             "org.freedesktop.systemd1",
+                             "/org/freedesktop/systemd1",
+                             "org.freedesktop.systemd1.Manager",
+                             "JobRemoved",
+                             systemd_job_removed, *data);
+  if (UNLIKELY (ret < 0))
+    return crun_make_error (err, -ret, "sd-bus match signal");
+
+  return 0;
+}
+
+static int
+systemd_check_job_status (sd_bus *bus,
+                          void *data,
+                          const char *path,
+                          const char *op,
+                          libcrun_error_t *err)
+{
+  int sd_err;
+  cleanup_free struct systemd_job_removed_s *d = data;
+
+  d->path = path;
+  while (!d->terminated)
+    {
+      sd_err = sd_bus_process (bus, NULL);
+      if (UNLIKELY (sd_err < 0))
+        return crun_make_error (err, -sd_err, "sd-bus process");
+
+      if (sd_err == 0)
+        {
+          sd_err = sd_bus_wait (bus, (uint64_t) -1);
+          if (UNLIKELY (sd_err < 0))
+            return crun_make_error (err, -sd_err, "sd-bus wait");
+
+          continue;
+        }
+    }
+
+  if (d->unit && d->result)
+    {
+      if (strcmp(d->result, "done") != 0)
+          return crun_make_error (err, 0, "error %s systemd unit %s: %s", op, d->unit, d->result);
+
+      free(d->unit);
+      free(d->result);
```

Comment:
these won't be freed if the function returns earlier

Reply:
I think it will easier if the ownership for data is owned by the caller (so the cleanup can happen in the `exit:` section), and free both the strings and the struct itselct could be on the stack as it is really small

Reply:
Are you suggesting moving the code that checks and generates the error into the callback function (`systemd_job_removed`)? Let me see...

Reply:
It looks like the only way to allocate the `struct systemd_job_removed_s` on the stack would be to merge `systemd_check_job_status_setup` into `systemd_check_job_status`.

I don't see any way to allocate the `unit` and `result` fields on the stack since `systemd_job_removed` is a callback. The only way to not allocate those are to remove those (moving the error generation into `systemd_job_removed` (which might not be a bad idea).